### PR TITLE
Better non-owning any implementation

### PR DESCRIPTION
**Pull request by lisitsyn**

### CODE REVIEW

File: src/shogun/lib/any.h [hunk 1]
```diff
@@ -207,7 +197,7 @@ namespace shogun
 		 */
 		virtual void set(void** storage, const void* v) const
 		{
-			*(storage) = const_cast<void*>(v);
+			*static_cast<T*>(*(storage)) = T(*reinterpret_cast<T const*>(v));
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: party

File: src/shogun/lib/any.h [hunk 2]
```diff
@@ -324,11 +312,20 @@ namespace shogun
...
+				return *(this);
+			}
 			policy->clear(&storage);
-			policy = other.policy;
-			policy->set(&storage, other.storage);
-			return *(this);
+			if (other.policy->policy_type() == PolicyType::NON_OWNING) {
+				policy = other.policy;
+				storage = other.storage;
+				return *(this);
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: could be moved outside the `if/then/else`

File: tests/unit/lib/Any_unittest.cc [hunk 3]
```diff
@@ -105,13 +105,94 @@ TEST(Any, erase_type_non_owning)
...
+	int32_t integer = 10;
+	int32_t other = 42;
+	auto any = erase_type_non_owning(&integer);
+	EXPECT_EQ(any.as<int32_t>(), integer);
+	any = erase_type(other);
+	EXPECT_EQ(any.as<int32_t>(), other);
+	EXPECT_EQ(integer, other);
+}
+
+TEST(Any, assign_non_owning_into_non_owning_then_owning)
```
Comment (karlnapf) [inline tests/unit/lib/Any_unittest.cc]: :D

File: src/shogun/lib/any.h [hunk 4]
```diff
@@ -396,28 +393,19 @@ namespace shogun
 			return policy->type_info();
 		}
 
-	private:
-		void assert_same_policy_type(BaseAnyPolicy* other_policy)
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: I understand we don't need this anymore? Is that why it is removed?

**Comment by karlnapf** [comment]
:heart:

File: src/shogun/lib/any.h (hunk 4 above)
Comment (lisitsyn) [reply src/shogun/lib/any.h]: Yes, but we need another check now

**Comment by lisitsyn** [comment]
I believe it should cover all of them. But I missed the type-checking part so need to add a few more tests and checks.

File: src/shogun/lib/any.h (hunk 1 above)
Comment (lisitsyn) [reply src/shogun/lib/any.h]: Funny is that I don't think it actually was correct before.

**Comment by lisitsyn** [comment]
Some SGObject tests are failing yet.

**Comment by lisitsyn** [comment]
@karlnapf all green, could you please take a look again and merge if all good

**Pull request merged**